### DISCUSSION TITLE
Reconciliation sensor behaviour on partitioned assets

### DISCUSSION THREAD

**Discussion by pyk** [dagster-io/dagster#12341] [Q&A] [answered]
I have a few questions about the behaviour of reconciliation sensor on partitioned assets.

Suppose I have the following partitioned assets:

```python

example_daily = dagster.DailyPartitionsDefinition(
    start_date="2023-02-01", end_offset=1
)


@dagster.asset(partitions_def=example_daily)
def example_upstream(context: dagster.OpExecutionContext) -> list[int]:
    return [1, 2, 3]


@dagster.asset(
    partitions_def=example_daily,
    ins={
        "persisted_list": dagster.AssetIn("example_upstream"),
    },
)
def example_downstream_1(
    context: dagster.OpExecutionContext, persisted_list: list[int]
) -> list[int]:
    return persisted_list * 2


@dagster.asset(
    partitions_def=example_daily,
    ins={
        "persisted_list_1": dagster.AssetIn("example_upstream"),
        "persisted_list_2": dagster.AssetIn("example_downstream_1"),
    },
)
def example_downstream_2(
    context: dagster.OpExecutionContext,
    persisted_list_1: list[int],
    persisted_list_2: list[int],
) -> int:
    return len(persisted_list_1) + len(persisted_list_2)
```

I want to keep `example_downstream_1` and `example_downstream_2` up-to-date if `example_upstream` is updated, regardless the partition key. 

So I create the following sensor:

```python
example_reconciliation_sensor = dagster.build_asset_reconciliation_sensor(
    name="example_reconciliation_sensor",
    asset_selection=dagster.AssetSelection.keys("example_upstream").downstream(
        include_self=False
    ),
)
```

If I run the materialization of `example_upstream` with today partition `2023-02-14`, the sensor materialize `example_downstream_1` and `example_downstream_2` assets as expected:

<img width="1360" alt="Screenshot 2023-02-15 at 05 00 28" src="https://user-images.githubusercontent.com/2213646/218873142-6af17a1e-38d6-431b-822c-e090c931e96a.png">

But if I run the materialization of `example_upstream` with yesterday partition `2023-02-13`, the sensor doesn't materialize `example_downstream_1` and `example_downstream_2` assets.

My questions:
1. Is this behaviour expected?
2. If this is expected, how to make reconciliation sensor to work with oldest/previous partition?

**Comment by OwenKephart** [comment, marked as answer]
Hi @pyk! Good question. The short answer is that this is expected behavior in the short term, but that making it possible to customize this behavior is high-priority in the medium-term. There's some more detailed discussion on this topic here: https://github.com/dagster-io/dagster/discussions/11829#discussioncomment-4752024

The gist of it is that if, by default, the reconciliation sensor were to act on arbitrarily old partitions of an asset, then adding a new asset with a large history of partitions which have not yet been materialized would result in the undesirable and surprising behavior of the sensor attempting to materialize all of those missing partitions immediately.

**Reply by pyk** [reply]
@OwenKephart hi thank you for the explanation it helps me!

> The short answer is that this is expected behavior in the short term, but that making it possible to customize this behavior is high-priority in the medium-term.

Does this statement true if freshness policy is enabled for the downstream asset? I assume the freshness policy is working per `partition_key` (?)
